### PR TITLE
[Core] Fix placement group leaks

### LINKED ISSUE #26761: [Core] negative resources
### What happened + What you expected to happen

```python
Healthy:
 1 ray.head.default
 1 ray.worker.default
Pending:
 (no pending nodes)
Recent failures:
 (no failures)

Resources
---------------------------------------------------------------
Usage:
 3.0/8.0 CPU (3.0 used of 3.0 reserved in placement groups)
 2.0/2.0 GPU (2.0 used of 2.0 reserved in placement groups)
 -8.0/2.0 accelerator_type:T4 (0.0 used of 10.0 reserved in placement groups)
 0.00/20.030 GiB memory
 0.05/8.895 GiB object_store_memory

Demands:
 {'CPU': 1.0, 'node:10.0.2.102': 0.001}: 1+ pending tasks/actors (1+ using placement groups)
 {'CPU': 1.0, 'node:10.0.2.51': 0.001}: 1+ pending tasks/actors (1+ using placement groups)
```

Hi @rkooo567, i observed this again.

### Versions / Dependencies

ray 3.0.0

### Reproduction script

N/A

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Root cause: When we schedule a task, we allocate resources and start a worker. If cancel bundle request is received before a worker is started, there's a leak because cancel bundle kills a worker and returns resources from "workers that are already started". 

It fixes the issue by retrying cancellation. This also means

- If a worker starts late (it has 60 seconds timeout), retry can fail as we have max number of retry. We retry for a very long time (10 * registration timeout), so it is unlikely to happen

Alternatively, to improve the consistency, we can also do
- register removed pg and keep deleting resources (with a reconciler) until it is fully gone.
- register leased workers before workers are started. This can be a better solution, but the implication of this change is unknown. 

I chose the current solution as it is needed for a network issue as well.  

## Related issue number

Closes https://github.com/ray-project/ray/issues/26761

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rynewang** [comment]
can we have a unit test, e.g. start a pg and kill it right away, and see if there's any leak?

**Comment by rkooo567** [comment]
@rynewang I am figuring that out. This requires very subtle timing, so writing an unit test is not trivial. Writing cpp unit test is even harder because of how our code is structured.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc [hunk 1]
```diff
@@ -563,7 +589,14 @@ void GcsPlacementGroupScheduler::DestroyPlacementGroupCommittedBundleResources(
...
       auto &bundle_spec = iter.second.second;
       auto &node_id = iter.second.first;
-      CancelResourceReserve(bundle_spec, gcs_node_manager_.GetAliveNode(node_id));
+      CancelResourceReserve(
+          bundle_spec,
+          gcs_node_manager_.GetAliveNode(node_id),
+          // Retry 10 * worker registeration timeout to avoid race condition.
+          // See https://github.com/ray-project/ray/pull/42942
+          // for more details.
+          /*max_retry*/ RayConfig::instance().worker_register_timeout_seconds() * 10,
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:598]: Q: should we do exponential backoff? Right now, we retry every 1 second.

**Comment by rkooo567** [comment]
@jjyao it is ready to be reviewed

**Comment by jjyao** [comment]
> There could be transient state where negative resources can exist.

When can it happen?

**Comment by rkooo567** [comment]
> There could be transient state where negative resources can exist.

When a worker is not started yet!

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc [hunk 2]
```diff
@@ -241,10 +245,25 @@ void GcsPlacementGroupScheduler::CancelResourceReserve(
 
   return_client->CancelResourceReserve(
       *bundle_spec,
-      [bundle_spec, node_id](const Status &status,
-                             const rpc::CancelResourceReserveReply &reply) {
-        RAY_LOG(DEBUG) << "Finished cancelling the resource reserved for bundle: "
-                       << bundle_spec->DebugString() << " at node " << node_id;
+      [this, bundle_spec, node_id, node, max_retry, current_retry_cnt](
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:248]: Where do you use the max retry

File: src/ray/raylet/scheduling/cluster_resource_manager.cc [hunk 3]
```diff
@@ -59,8 +59,6 @@ void ClusterResourceManager::AddOrUpdateNode(
 
 void ClusterResourceManager::AddOrUpdateNode(scheduling::NodeID node_id,
                                              const NodeResources &node_resources) {
-  RAY_LOG(DEBUG) << "Update node info, node_id: " << node_id.ToInt()
```
Comment (jjyao) [inline src/ray/raylet/scheduling/cluster_resource_manager.cc:62]: Why remove?

Comment (rkooo567) [reply src/ray/raylet/scheduling/cluster_resource_manager.cc:62]: This seems useless and too verbose. I can bring it back if you think it is necessary

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:248]: oops.. let me fix this. I will also add an unit test

**Comment by rkooo567** [comment]
@jjyao it'd be great if you can approve the PR. I will add changes and merge it

**Review by jjyao — APPROVED** [review]

**Pull request merged**